src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+                    if seen[j] {
+                        continue;
+                    }
+
+                    // Are the two selectors excluded from co-existing?
+                    // j > i and so (j, i) is always on the lower diagonal
+                    if exclusion_matrix[j][i] {
+                        continue;
+                    }
+
+                    // Can the new selector join the combination?
+                    if degrees[j] + usage <= max_degree {
+                        // Yes, it can. Let's do it.
+                        combination.push(j);
+                        // The remaining room available is the lesser of the
+                        // current remaining room available for previous
+                        // selector(s) after adjusting for adding this new
+                        // selector, and the room available in the new selector
+                        // after adjusting for adding the previous selectors.
+                        room = std::cmp::min(room - 1, max_degree - (degrees[j] + usage));
+                        usage += 1;
+                        selector_map[j] = combinations.len();
+                        seen[j] = true;
+                    }
+                }
+
+                combinations.push(combination);
+            }
+        }
+
+        let mut selector_replacements = vec![None; self.num_selectors];
+        let mut polys = vec![vec![F::zero(); n]; combinations.len()];
+        let mut new_columns = vec![];
+        for (combination, poly) in combinations.iter().zip(polys.iter_mut()) {
+            // Create a new fixed column for the logical selector
+            let column = self.fixed_column();
+            new_columns.push(Some(column));
+            let query = Expression::Fixed {
+                query_index: self.query_fixed_index(column, Rotation::cur()),
+                column_index: column.index,
+                rotation: Rotation::cur(),
+            };
+
+            // Every virtual selector gets a different integer (starting at 1)
+            // that it assigns when it wants to be active.
+            let mut assignment = F::one();
+            for &selector in combination {
+                for (s, poly) in selectors[selector].iter().zip(poly.iter_mut()) {
+                    if *s {
+                        *poly = assignment;
+                    }
+                }
+                // Construct a polynomial that has a root at 0 and at integers
+                // not including what this selector has been assigned to. This
+                // is a generalization of a boolean selector when the selector
+                // is multiplying a value that is constrained to equal zero.
+                let mut expr = query.clone();
+                let mut root = F::one();
+                for _ in 0..combination.len() {
+                    if root != assignment {
+                        expr = expr * (query.clone() - Expression::Constant(root));
```

Comment:
```suggestion
                        expr = expr * (Expression::Constant(root) - query.clone());
```
per https://github.com/zcash/orchard/issues/143.